### ISSUE TITLE
AWS.SharedIniFileCredentials is not a constructor

### ISSUE THREAD

**Issue by linshu** [aws/aws-sdk-js#1286]
I got this error:

```
__WEBPACK_IMPORTED_MODULE_1_aws_sdk__.SharedIniFileCredentials is not a constructor
```

 while I call like this function in typescript like this:

```
let cred = new AWS.SharedIniFileCredentials({profile: 'default'});
```

**Comment by chrisradek** [comment]
@linshu 
Can you share some more information on how you're using the SDK?

Are you using the browser version of the SDK, or are you using webpack to generate a node bundle? The browser version of the SDK doesn't support `SharedIniFileCredentials` since it requires file system access. 

If you're trying to create a node bundle, can you share your webpack config?

**Comment by linshu** [comment]
@chrisradek 

I use a project generated from angular-cli. Other aws apis seem work fine. The original error is like this:

```
error_handler.js:52 ORIGINAL EXCEPTION: __WEBPACK_IMPORTED_MODULE_1_aws_sdk__.SharedIniFileCredentials is not a constructor
ErrorHandler.handleError @ error_handler.js:52
next @ application_ref.js:346
schedulerFn @ async.js:91
SafeSubscriber.__tryOrUnsub @ Subscriber.js:223
SafeSubscriber.next @ Subscriber.js:172
Subscriber._next @ Subscriber.js:125
Subscriber.next @ Subscriber.js:89
Subject.next @ Subject.js:55
EventEmitter.emit @ async.js:77
NgZone.triggerError @ ng_zone.js:329
onHandleError @ ng_zone.js:290
ZoneDelegate.handleError @ zone.js:246
Zone.runTask @ zone.js:154
ZoneTask.invoke @ zone.js:345
```

The code is like this:
```
import { Injectable } from "@angular/core";
import * as AWS from 'aws-sdk';

@Injectable()
export class AwsService {
    loggedin: boolean;
    aws_accessKeyId: string;
    aws_secretAccessKey: string;
    aws_region: string;

    constructor() {
        this.loggedin = false;
        this.aws_region = 'us-west-2';
    }

    login(
    ) {
        let cred = new AWS.SharedIniFileCredentials({profile: 'default'});
        AWS.config.update({
            region: this.aws_region,
            credentials: {
                accessKeyId: this.aws_accessKeyId,
                secretAccessKey: this.aws_secretAccessKey
            }
        });
        this.loggedin = true;
        console.log(this.aws_accessKeyId, this.aws_secretAccessKey);
    }
```

The package.json:

```
{
  "name": "crossbow",
  "version": "1.0.0",
  "scripts": {
    "ng": "ng",
    "start": "ng serve",
    "lint": "tslint \"src/**/*.ts\"",
    "test": "ng test",
    "pree2e": "webdriver-manager update --standalone false --gecko false",
    "e2e": "protractor"
  },
  "licenses": [
    {
      "type": "MIT",
      "url": "https://github.com/angular/angular.io/blob/master/LICENSE"
    }
  ],
  "angular-cli": {},
  "private": true,
  "dependencies": {
    "@angular/common": "^2.3.1",
    "@angular/compiler": "^2.3.1",
    "@angular/core": "^2.3.1",
    "@angular/forms": "^2.3.1",
    "@angular/http": "^2.3.1",
    "@angular/platform-browser": "^2.3.1",
    "@angular/platform-browser-dynamic": "^2.3.1",
    "@angular/router": "^3.3.1",
    "angular2-uuid": "^1.1.0",
    "aws-sdk": "^2.7.11",
    "bootstrap": "^3.3.7",
    "core-js": "^2.4.1",
    "d3": "^4.4.0",
    "ng2-bootstrap": "^1.1.16-11",
    "rxjs": "^5.0.1",
    "ts-helpers": "^1.1.1",
    "zone.js": "^0.7.2"
  },
  "devDependencies": {
    "@angular/compiler-cli": "^2.3.1",
    "@types/aws-sdk": "0.0.42",
    "@types/jasmine": "2.5.38",
    "@types/node": "^6.0.55",
    "angular-cli": "1.0.0-beta.24",
    "codelyzer": "~2.0.0-beta.1",
    "jasmine-core": "2.5.2",
    "jasmine-spec-reporter": "2.5.0",
    "karma": "1.2.0",
    "karma-chrome-launcher": "^2.0.0",
    "karma-cli": "^1.0.1",
    "karma-jasmine": "^1.0.2",
    "karma-remap-istanbul": "^0.2.1",
    "protractor": "~4.0.13",
    "ts-node": "1.2.1",
    "tslint": "^4.0.2",
    "typescript": "~2.0.3"
  }
}
```

tsconfig.json:

```
{
  "compilerOptions": {
    "baseUrl": "",
    "declaration": false,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "lib": ["es5", "dom"],
    "mapRoot": "./",
    "module": "es6",
    "moduleResolution": "node",
    "outDir": "../dist/out-tsc",
    "sourceMap": true,
    "target": "es5",
    "typeRoots": [
      "../node_modules/@types"
    ]
  }
}
```

**Comment by chrisradek** [comment]
@linshu 
Is this meant to be run in the browser? That's what it looks like to me but correct me if I'm wrong. If the SDK is running in the browser, it won't have access to `AWS.SharedIniFileCredentials`, since that is meant to source credentials from the file system, which the SDK won't have access to in the browser. You would need to use a different credential provider (such as `CognitoIdentityCredentials`) instead.

**Comment by linshu** [comment]
@chrisradek Thanks for your answer. I changed my code like below:

```
import { Injectable } from "@angular/core";
import * as AWS from 'aws-sdk';

@Injectable()
export class AwsService {
    loggedin: boolean;
    aws_accessKeyId: string;
    aws_secretAccessKey: string;
    aws_region: string;

    constructor() {
        this.loggedin = false;
        this.aws_region = 'us-west-2';
    }

    login(
    ) {
        let cred = new AWS.CognitoIdentityCredentials({
            IdentityPoolId: 'us-west-2_xxxxxx'
        });
        AWS.config.update({
            region: this.aws_region,
            credentials: cred
        });
        this.loggedin = true;
        console.log(this.aws_accessKeyId, this.aws_secretAccessKey);
    }
}
```

But, I got this new error. The credential seems not working in config().

```
Error: Missing credentials in config
    at Request.extractError (json.js:43) [angular]
    at Request.callListeners (sequential_executor.js:105) [angular]
    at Request.emit (sequential_executor.js:77) [angular]
    at Request.emit (request.js:668) [angular]
    at Request.transition (request.js:22) [angular]
    at AcceptorStateMachine.runTo (state_machine.js:14) [angular]
    at :4200/vendor.bundle.js:98018:10 [angular]
    at Request.<anonymous> (request.js:38) [angular]
    at Request.<anonymous> (request.js:670) [angular]
    at Request.callListeners (sequential_executor.js:115) [angular]
    at Request.emit (sequential_executor.js:77) [angular]
    at Request.emit (request.js:668) [angular]
    at Request.transition (request.js:22) [angular]
    at AcceptorStateMachine.runTo (state_machine.js:14) [angular]
```

Also, I think this is not what I want. I just want my app to read a credential from some configuration without user having to input since my app is just a simple QA test app. I want to make it as simple as possible. Is there any other better way I can use?

Thanks a lot.

**Comment by chrisradek** [comment]
@linshu 
Apologies, I added a fix for this with version [`2.7.28`](https://github.com/aws/aws-sdk-js/blob/master/CHANGELOG.md#2728) of the SDK but didn't report it here.

The reason you were seeing `Missing credentials in config` when using `CognitoIdentityCredentials` was because there wasn't a region configured. The only way to set the region would have been to set it on the `AWS.config` object before instantiating your credentials.

Now, you can specify a `region` to use in the `CognitoIdentityCredentials` constructor:
https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CognitoIdentityCredentials.html#constructor-property
